xla/service/gpu/runtime/nccl_collective_permute_thunk.cc

```diff
@@ -186,24 +246,46 @@ absl::Status RunCollectivePermute(
                                 device_string, current_id,
                                 source_id.value_or(-1), target_id.value_or(-1));
 
+  // If sending to another peer, get the pointer value of the src addr.
+  // Only change the pointer value when it's different from stored one.
+  if (target_id && send_ptr_value != (uint64_t)src_addr.opaque()) {
+    send_ptr_value = (uint64_t)src_addr.opaque();
+  }
   // GroupStart/End API is needed only if we will issue both send & recv calls.
   const bool is_nccl_group_needed = (target_id && source_id);
   if (is_nccl_group_needed) {
     TF_RETURN_IF_ERROR(nccl_api->GroupStart());
   }
 
-  // Send source buffer to target peer if needed.
-  if (target_id) {
-    TF_RETURN_IF_ERROR(nccl_api->Send(src_addr, buffer.element_type,
-                                      buffer.element_count, *target_id, comm,
-                                      &stream));
-  }
-
-  // Receive data from the source peer to the destination buffer.
-  if (source_id) {
-    TF_RETURN_IF_ERROR(nccl_api->Recv(dest_addr, buffer.element_type,
-                                      buffer.element_count, *source_id, comm,
-                                      &stream));
+  // If all peers are local, only get/send device pointer values and invoke
+  // memcpy.
+  if (use_memcpy) {
```

Comment:
I think you should be able to do this without any calls to NCCL, if all peers are local you can use `RendezvousOnce` (it also has a std::once_flag like construct to make sure it's executed once) to exchange pointers between all thunks participating in the send/recv, and  then immediately to memcpy

Reply:
You don't even need rendezvous, you should be able to simply put it into a flat_hash_map? Add `AsyncValue<void*>` when requestion a pointer, and make it available when "publishing" a pointer?

Reply:
Ok that AsyncValue approach really simplifies the code quite a bit.